ethcore/src/engines/authority_round/finality.rs

```diff
@@ -110,7 +111,14 @@ impl RollingFinality {
 	/// Returns a list of all newly finalized headers.
 	// TODO: optimize with smallvec.
 	pub fn push_hash(&mut self, head: H256, signers: Vec<Address>) -> Result<Vec<H256>, UnknownValidator> {
-		if signers.iter().any(|s| !self.signers.contains(s)) { return Err(UnknownValidator) }
+		// TODO: seems bad to iterate over signers twice like this.
```

Comment:
It's alright. There will usually be only one element in the list (block author), there might be a bit more if there are empty steps.